[Compose 페이먼츠 4단계] 미플 미션 제출합니다

HamBeomJoon opened the pull request:
# 셀프 체크리스트

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?
- 데모데이때문에 시간없어서 코드가 ㅎㅎ 😂

## 스크린샷

https://github.com/user-attachments/assets/9581bad4-70f6-4b4a-9621-e44529f06191

HamBeomJoon committed:
docs: 4단계 기능 요구사항 정의

HamBeomJoon committed:
refactor: `backgroundColor` 속성 제거 및 `Color` 타입 Parcelize 처리

- `BankUiModel`의 `backgroundColor` 속성을 제거하고 `color` 속성의 타입을 `Int`에서 `Color`로 변경
- `Color` 타입의 `Parcelable` 처리를 위해 `ColorParceler` 추가

HamBeomJoon committed:
feat: 은행 선택 BottomSheet 숨김 처리 개선

은행 선택 시 BottomSheet가 먼저 숨겨지고 은행 정보가 업데이트 되도록 수정했습니다.

HamBeomJoon committed:
feat: 카드사 선택 상태를 관리하는 CardCompanySelectionState 추가

- `AddCardUiState`의 카드사 관련 로직을 `CardCompanySelectionState`를 사용하도록 수정
- `BankSelectBottomSheet`에서 `BankType.NOT_SELECTED`를 필터링하지 않도록 변경하고, `BankUiModel.NOT_SELECTED` 추가

HamBeomJoon committed:
feat: column -> LazyColumn으로 변경

HamBeomJoon committed:
fix: LazyColumn 버그 수정

HamBeomJoon committed:
feat: 카드 목록 화면에서 카드 수정 기능 구현

- 카드 목록 화면에서 기존 카드를 클릭하면 카드 수정 화면으로 이동하도록 구현
- `navigateToAddCard` 함수의 매개변수에 `cardToEdit` 추가하여 카드 정보를 전달

HamBeomJoon committed:
feat: 카드 수정 기능 추가

HamBeomJoon committed:
feat: 카드 수정 기능 추가

카드 수정 기능을 위해 `CardListStateHolder`의 `addCard` 함수를 `upsertCard`로 변경했습니다.
`CardListActivity`에서 카드를 수정할 수 있도록 `navigateToAddCard` 함수에 `cardToEdit` 파라미터를 추가하고,
`EDIT_CARD_KEY`를 정의했습니다.

HamBeomJoon committed:
test: UI Test 및 contentDescription 수정

- UI 테스트의 contentDescription을 한글로 수정
- UI 테스트 fixture에 id 추가
- BankItem의 contentDescription에서 은행 문구 제거

HamBeomJoon committed:
fix: 첫 번째 카드 수정 시 새로 카드 추가되는 현상 수정

HamBeomJoon committed:
feat: 카드 수정 사항 없으면 토스트 메시지 표시

- AddCardUiState에 카드 회사 선택 상태 추가
- AddCardScreen에 카드 수정/추가 모드에 따라 TopBar 제목 변경
- NewCardTopBar에 title 파라미터 추가
- AddCardStateHolder에 수정 모드 및 변경 사항 확인 로직 추가
- BankUiModel에 도메인 모델로 변환하는 함수 추가

HamBeomJoon committed:
feat: 카드 수정 시 토스트 메시지 분기 처리

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/NewCardTopBar.kt:
```diff
@@ -51,6 +50,7 @@ fun NewCardTopBar(
 fun CardTopBarPreview() {
```
Preview를 Public으로 설정하신 이유가 없다면 Private로 하면 좋을 것 같아요

moondev03 commented on app/src/main/java/woowacourse/payments/ui/model/BankUiModel.kt:
```diff
@@ -1,21 +1,22 @@
-    val color: Int,
-    val logoRes: Int,
+    @TypeParceler<Color, ColorParceler> val color: Color,
+    @param:DrawableRes val logoRes: Int,
```
`@param:DrawableRes`와 `@DrawableRes`는 어떤 차이가 있나요?

HamBeomJoon replied:
저는 @DrawableRes를 쓰면 자꾸 경고가 뜨더라고요..

> This annotation is currently applied to the value parameter only, but in the future it will also be applied to 'field'. See https://youtrack.jetbrains.com/issue/KT-73255 for more details. To remove this warning, use the '@param:' annotation target.

> 지금 코틀린 버전에서는 @DrawableRes가 생성자 파라미터(param)에만 적용됨
하지만, 미래 코틀린 버전에서는 field에도 자동 적용될 예정

Parcelize, 리플렉션, IDE/검증 로직 등에서 field 존재 여부에 따라 동작이 달라질 수 있어서 명시적으로 @param을 붙이라고 경고가 뜨는 것 같습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/model/CardUiModel.kt:
```diff
@@ -5,6 +5,7 @@ import kotlinx.parcelize.Parcelize
 @Parcelize
 data class CardUiModel(
+    val id: Long,
     val bankUiModel: BankUiModel,
```
bankUiModel 보다 bank는 어떠신가요?

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardActivity.kt:
```diff
@@ -6,29 +6,38 @@ import androidx.activity.ComponentActivity
         enableEdgeToEdge()
+        val cardToEdit: CardUiModel? = intent.getParcelableExtraCompat(EDIT_CARD_KEY)
         setContent {
             AndroidpaymentsTheme {
                 val stateHolder =
-                    rememberSaveable(saver = AddCardStateHolder.saver) { AddCardStateHolder() }
+                    rememberSaveable(saver = AddCardStateHolder.saver) {
+                        AddCardStateHolder(cardToEdit?.toAddCardUiState() ?: AddCardUiState())
+                    }
 
                 AddCardScreen(
+                    stateHolder = stateHolder,
                     onBackPressed = { finish() },
                     onCardSaved = { cardUiModel ->
-                        val resultIntent =
-                            Intent().apply {
-                                putExtra(NEW_CARD_KEY, cardUiModel)
-                            }
+                        val cardWithId =
+                            cardToEdit?.let {
+                                cardUiModel.copy(id = it.id)
+                            } ?: cardUiModel.copy(
+                                id = 0L,
+                            )
+                        val resultIntent = Intent().apply { putExtra(NEW_CARD_KEY, cardWithId) }
                         setResult(RESULT_OK, resultIntent)
                         finish()
                     },
-                    stateHolder = stateHolder,
                 )
             }
         }
```
이번 강의에서 다룬 `재사용 vs 중복` 주제와 관련해,
미플은 카드 수정과 등록을 하나의 화면에서 모두 처리하도록 구현하신 것으로 보였습니다.

이 방식을 선택하신 이유가 궁금합니다!

HamBeomJoon replied:
맨위에 topBar 텍스트말고는 모두 같아서 재사용을 했습니다. (카드 등록화면에서도 저는 카드번호 등을 표시하고 있어서 😅)

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardScreen.kt:
```diff
@@ -54,10 +59,15 @@ fun AddCardScreen(
         modifier = Modifier.fillMaxSize(),
         topBar = {
             NewCardTopBar(
+                title = if (stateHolder.isEditMode) "카드 수정" else "카드 추가",
```
하드 코딩 되어있네요 🤔

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardScreen.kt:
```diff
@@ -54,10 +59,15 @@ fun AddCardScreen(
                 onBackClick = onBackPressed,
                 onSaveClick = {
+                    if (!stateHolder.hasChanges()) {
+                        Toast.makeText(context, "변경된 사항이 없습니다", Toast.LENGTH_SHORT).show()
```
하드 코딩 되어있네요 🤔

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardScreen.kt:
```diff
@@ -54,10 +59,15 @@ fun AddCardScreen(
+                title = if (stateHolder.isEditMode) "카드 수정" else "카드 추가",
                 onBackClick = onBackPressed,
                 onSaveClick = {
+                    if (!stateHolder.hasChanges()) {
```
버튼 클릭 시 메시지를 노출하는 방법과 아예 버튼을 비활성화하는 방법 중 전자를 선택하신 이유가 궁금해요.

HamBeomJoon replied:
그냥 사용자에게 명확하게 알림?을 주고자 토스트 메세지를 노출했습니다

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardStateHolder.kt:
```diff
@@ -4,18 +4,23 @@ import androidx.compose.runtime.getValue
     initialState: AddCardUiState = AddCardUiState(),
+    private val originalState: AddCardUiState = initialState,
 ) {
+    val allBanks: List<BankType> = BankType.entries
```
BankUiModel을 생성한 걸로 확인했는데 외부에서 BankUiModel을 사용하지 않고 BankType 도메인 모델을 사용하신 이유가 있으신가요?

HamBeomJoon replied:
UiModel을 만들었고, AddCardStateHolder가 결국 ui(presentation) 레이어니까 해당 모델을 사용하는 것이 맞을 것 같네요. 수정했습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardStateHolder.kt:
```diff
@@ -4,18 +4,23 @@ import androidx.compose.runtime.getValue
     var uiState by mutableStateOf(initialState)
         private set
 
+    val isEditMode: Boolean = originalState != AddCardUiState()
```
`isEditMode`와 `hasChanges()`는 각각 어떤 역할을 하고 있나요?

HamBeomJoon replied:
isEditMode는 카드 추가 페이지에 들어갈때 추가상태인지 수정 상태인지 구분하기 위해 추가했습니다.
hasChanges()는 카드 수정모드에 들어간 뒤 변화가 없으면 저장못하게 한다? 그런 요구사항때문에 추가했습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/CardCompanySelectionState.kt:
```diff
@@ -0,0 +1,14 @@
+import kotlinx.parcelize.Parcelize
+import woowacourse.payments.domain.BankType
+
+@Parcelize
+sealed class CardCompanySelectionState : Parcelable {
+    data object NotSelected : CardCompanySelectionState()
+
+    data class Selected(
+        val bank: BankType,
+    ) : CardCompanySelectionState()
+}
```
별도의 State를 통해 카드 선택 여부를 확인하신 이유가 궁금합니다

HamBeomJoon replied:
저번에 NOT_SELECTED가 도메인이 아닌 것 같다해서 nullable하게 두는 방법과 위처럼 하는 방식 2가지가 떠올랐는데, 한번 sealed class를 사용해보고 싶어서 선택했습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListScreen.kt:
```diff
@@ -42,28 +42,33 @@ import woowacourse.payments.ui.theme.AndroidpaymentsTheme
 fun CardListScreen(
     stateHolder: CardListStateHolder,
-    navigateToAddCard: () -> Unit,
+    navigateToAddCard: (cardToEdit: CardUiModel?) -> Unit,
```
이름은 `navigateToAddCard`인데 실제로는 생성과 수정 모두 처리하도록 하신 것 같아요.
하나로 처리하신 이유가 무엇인가요?

HamBeomJoon replied:
일단 시간없어서... 네이밍은 신경안썼습니다 🥲

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardScreen.kt:
```diff
@@ -114,9 +124,12 @@ fun AddCardScreen(
     if (showBottomSheetState) {
         BankSelectBottomSheet(
```
<img width="267" height="593" alt="Image" src="https://github.com/user-attachments/assets/61ca238d-1b28-48bb-94f1-575359633e61" />

현재는 바텀시트가 한 번 내려가면, 카드를 선택해도 다시 열리지 않아 사용자가 화면을 나갔다가 다시 들어와야 하는 불편함이 있는 것 같아요.

또, 처음에 은행을 선택했더라도 다른 은행으로 변경하고 싶은 경우 바텀시트가 다시 올라오지 않아 동일한 문제가 발생할 수 있습니다.

카드를 클릭했을 때 바텀시트가 다시 열리도록 해보는 건 어떨까요?

HamBeomJoon replied:
하긴해야했는데, 시간이 없어서 못하다 이제 했네요 ㅋㅋ. 카드 클릭 시 바텀시트가 열리도록 변경했습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListStateHolder.kt:
```diff
@@ -9,6 +9,7 @@ import woowacourse.payments.ui.model.CardUiModel
 class CardListStateHolder(
     initialCards: List<CardUiModel> = emptyList(),
```
> AddCardStateHolder에서는 uiState를 매개변수로 받았는데, 여기서는 initialCards를 받아 내부에서 uiState를 생성하고 있네요. 어떤 차이가 있나요?

저번 리뷰에서 위와 같은 질문을 드렸는데, 답변이 실제 코드와는 조금 다른 부분이 있었던 것 같아요.

> CardListStateHolder는 List만 가지고 UI가 필요할 때마다 목록의 크기를 확인하여 showAddButton 값을 계산하면 되기 때문

이렇게 답변을 해주셨는데 실제로는 인자로 넣어준 `initialCards`를 보관하고 사용하는 것이 아닌 `UiState`를 생성하는데 사용되고 있습니다.

StateHolder마다 다르게 처리하신 이유나 의도가 궁금합니다.

---

추가로 `showAddButton`이라는 상태를 `initialCards`를 통해 초기에만 설정하고 있습니다.
다만, 만약 카드 삭제 기능이 추가되어 카드 삭제가 이루어진다면 지금은 AddButton이 사라지지 않고 계속 유지될 것 같아요.

차라리 UiState의 내부에 showAddButton을 계산하도록 하면 이와 같은 문제가 발생하지 않을 것 같습니다.

HamBeomJoon replied:
uiState 내부에 계산로직을 넣었습니다!

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListScreen.kt:
```diff
@@ -42,28 +42,33 @@ import woowacourse.payments.ui.theme.AndroidpaymentsTheme
         } else {
             CardListContent(
                 cards = uiState.cards,
                 enableScroll = uiState.cards.size > 1,
```
`uiState.cards.size > 1` vs `uiState.enableScroll`

HamBeomJoon replied:
uiState에서 계산된 프로퍼티로 수정했습니다!

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListStateHolder.kt:
```diff
@@ -17,13 +18,27 @@ class CardListStateHolder(
         uiState =
             uiState.copy(
                 cards = newCards,
                 showAddButton = newCards.size > 1,
```
위에서 말씀드린 내용과 동일하게 이렇게 매번 이렇게 동일한 로직을 계산해야 할까요?

HamBeomJoon replied:
uiState에서 계산된 프로퍼티로 수정했습니다!

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardUiState.kt:
```diff
@@ -37,8 +37,22 @@ data class AddCardUiState(
+        expired = this.expired,
+        cardOwner = this.owner,
+        password = "****",
+        cardCompanySelectionState = CardCompanySelectionState.Selected(bankUiModel.toDomain()),
```
`CardCompanySelectionState.Selected(bankUiModel.toDomain()),`

Domain으로 변경하는 이유가 무엇인가요?

HamBeomJoon replied:
AddCardUiState도 ui(presentation) 레이어니까 해당 모델을 사용하는 것이 맞을 것 같네요. 수정했습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListScreen.kt:
Preview 가시성

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardScreen.kt:
Preview 가시성

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardStateHolder.kt:
```diff
@@ -32,8 +37,9 @@ class AddCardStateHolder(
         uiState = uiState.copy(password = newPassword)
     }
```
현재는 카드 정보를 각각 다른 메서드를 통해 업데이트하고 있는데, 
그렇다면 각 항목에 대한 검증도 즉시 수행할 수 있지 않을까요?

지금처럼 등록 버튼을 눌렀을 때만 잘못된 값을 알려주는 방식보다는,
사용자가 입력하는 순간 빠르게 피드백을 주는 편이 더 나은 경험을 줄 수 있을 것 같습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardUiState.kt:
```diff
@@ -14,7 +14,7 @@ data class AddCardUiState(
     val cardOwner: String = "",
     val password: String = "",
-    val bankUiModel: BankUiModel = BankType.NOT_SELECTED.toPresentation(),
+    val cardCompanySelectionState: CardCompanySelectionState = CardCompanySelectionState.NotSelected,
```
CardCompanySelectionState를 별도로 두신 이유나, 이를 통해 얻은 장점이 어떤 건지 궁금합니다.

HamBeomJoon replied:
위에 설명했듯이, NOT_SELECTED를 도메인 모델에서 분리하는 과정에 있어서 해당 구조를 택했습니다. uiState처럼 카드사 선택여부를 state로 만들어 명시적 상태 관리와 나중에 추가로 state가 추가될 때도 대응할 수 있을 것 같았습니다.

moondev03 reviewed (changes requested):
미플 4단계 구현 고생하셨습니다!

궁금한 점들을 리뷰 내용으로 남겨두었으니 확인 후 답변 부탁드립니다. 😄

HamBeomJoon committed:
feat: preview 함수 private로 변경

HamBeomJoon committed:
refactor: 문자열 strings 분리

HamBeomJoon committed:
refactor: CardUiModel 매개변수 bankUiModel -> bank로 변경

HamBeomJoon committed:
refactor: CardListUiState의 상태 계산 로직 위임

`CardListUiState`가 카드 목록의 크기에 따라 `showAddButton`과 `enableScroll` 상태를 직접 계산하도록 변경했습니다. 이로 인해 StateHolder에서의 중복된 상태 업데이트 로직이 제거되었습니다.

HamBeomJoon committed:
refactor: 카드사 선택 관련 UI 모델 도메인 모델 사용 > BankUiModel 사용으로 변경

HamBeomJoon committed:
fix: 카드 선택 시 bottomSheet 표시되어 수정 가능하게 변경

HamBeomJoon committed:
chore: 불필요한 메서드 정리
